scraper/scraper.py

```diff
@@ -0,0 +1,212 @@
+import concurrent.futures
+import dotenv
+import itertools
+import json
+import logging
+import os
+import requests
+import sys
+import time
+
+from bs4 import BeautifulSoup
+
+# adjust recursion limit for parsing large files with Beautiful Soup
+sys.setrecursionlimit(10000) # default is 1000
+
+class CoursebookScraper:
+
+    def __init__(self, ptgsessid: str, congfiguration: dict = None, loggingFile: str = 'scraper.log'):
+        # handle object variables
+        self.ptgsessid = ptgsessid
+        self.configuration = congfiguration
+        self.loggingFile = loggingFile
+
+        # initialize logging for scraper
+        self.logger = logging.getLogger('Coursebook-Scraper')
+        logging.basicConfig(
+            format='%(asctime)s - %(levelname)s: %(message)s',
+            datefmt='%Y-%m-%d %H:%M:%S',
+            filename=self.loggingFile, 
+            filemode='w', 
+            level=logging.INFO, 
+        )
+
+        # document object instantiation
+        self.logger.info('Started coursebook scraper')
+
+
+    def getSearchSpecifications(self) -> dict:
+        self.logger.info('Obtaining CourseBook search specifications')
+        specifications = {}
+
+        # obtain coursebook index page
+        self.logger.info('Getting CourseBook index page')
+        try:
+            r = requests.get('https://coursebook.utdallas.edu')
+        except:
+            self.logger.error('Failed to obtain CourseBook index page')
+            return {}
+
+        # parse index page for search form
+        index = BeautifulSoup(r.content, 'html.parser')
+        try:
+            form = index.select_one('#guidedsearch')
+        except:
+            self.logger.error('Failed to parse CourseBook index page for search form')
+            return {}
+
+        # parse form fields for field name and options
+        try:
+            fields = form.find_all('tr')
+        except:
+            self.logger.error('Failed to parse search form for search fields')
+            return {}
+
+        for field in fields:
+            try:
+                title = field.select_one('th').text
+                options = [option['value'] for option in field.find_all('option')]
+                specifications[title] = options
+            except:
+                self.logger.error('Failed to parse a search field from search search form')
+
+        # return obtained specifications
+        return specifications
+
+
+    def downloadCoursebookData(self, queryList: list) -> dict:
+        # format query list as a proper query string
+        query = 'action=search'
+        for specifier in queryList:
+            query += '&s[]={0}'.format(specifier)
```

Comment:
```suggestion
            query += f'&s[]={specifier}'
```
Sidenote for future reference: I'm almost certain there's a library that does something like this, but this is fine.